CIPs/cip-0066.md

```diff
@@ -0,0 +1,113 @@
+
+## Motivation
+
+While CIP-66 txs will work like CIP-64 txs for end users in most cases, they provide multiple benefits.
+
+### Fee Accuracy
+Specifying the fees in CELO provides the same fee price accuracy for all fee currencies, even those with a low number of decimals. Some of the most popular ERC-20 tokens on Ethereum have only 6 decimals. That is enough to specify the total costs for a tx, but not sufficient to accurately specify the cost per unit of gas. E.g., for USDC, even the smallest possible `maxFeePerGas` of 0.00001 USDC without a tip would make a native transfer cost 21000 * 0.00001 = 0.21 USDC.
+
+### Increased Ethereum Compatibility
+
+The fields `maxPriorityFeePerGas`, `maxFeePerGas` and `effectiveGasPrice` are specified in the native CELO token, just like it is done for EIP-1559 txs. In many cases, this allows unmodified Ethereum tooling to work on these txs without knowing about CIP-66 and still getting the correct result. In contracts, CIP-64 txs would be misinterpreted by the exchange rate between CELO and the fee currency.
+Creating CIP-66 txs still requires explicit tooling support, though.
+
+### Simplified Data Analytics
+
+Data analysis queries, indexers and other tools often ask questions like "How much fees have been paid by account X", "What was the average tip?" or "What was the average cost of a tx?". When only looking at EIP-1559 (and to a certain degree legacy Ethereum-compatible txs) and CIP-66 txs, these questions can be answered without making any distinction between the different tx types, since they share the same fields *with the same meaning*. Although other tx types are in use now, this will likely change and make this a clear advantage in the long run.
+
+### Staying Closer to Upstream Code
+
+Keeping the tx field meaning consistent with EIP-1559 also allows us to reduce the necessary changes in celo-blockchain compared to geth. This is most relevant in two places:
+* The tx pool, where txs are sorted, replaced or discarded based on their fee values.
+* `baseFee` handling and `effectiveGasPrice` calculation (see [example diff](https://github.com/celo-org/op-geth/compare/palango/cleanup...celo-org:op-geth:palango/celo-denominated?expand=1))
+
+## Specification
+
+### Definitions
+
+- `TransactionType` `122` ⇒ `0x7a`. See [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
+- `||` is the byte/byte-array concatenation operator
+- `//` is integer division (discarding the remainder)
+- `feeCurrency` is the address of the whitelisted currency to be used to pay for gas. In contrast to older tx types, this field can not be omitted to mean "native CELO token".
+- The exchange rate from CELO to `feeCurrency` is specified as a rational number consisting of two `uint256`: `rateNum` / `rateDen`
+- `maxFeeInFeeCurrency` is chosen by the tx sender to limit the exchange rate risk. Txs must only be included if `maxFeeInFeeCurrency >= gasLimit * maxFeePerGas * rateNum // rateDen`
```

Comment:
so to send cip66 

a dev will need to set

`maxPriorityFeePerGas`, and `maxFeePerGas`, the same way as as any 1559 tx
`feeCurrency` same as cip64 tx

and then also `maxFeeInFeeCurrency`  -- are you expecting devs/ libraries to be fetching the exchange rates and calculating this number them selves?

Reply:
Unfortunately yes, and this is a certainly a downside of this proposal. If you want to set `maxFeeInFeeCurrency` sensibly (you could set it to maxInt and take all exchange rate risk), you'll have to do the equivalent of
```bash
cast call $(
	cast call 0x000000000000000000000000000000000000ce10 "getAddressForStringOrDie(string calldata identifier) returns (address)" SortedOracles
) 'function medianRate(address token) returns (uint256, uint256)' $TOKEN_ADDRESS
```
or get a reasonable exchange rate from a different source.